armv8 aarch64 EL3 init updates

sandeepbrcm opened the pull request:
1- Aarch64 EL3 init.  Initialize few recommended and few necessary settings of control registers before switching to EL1 ( arm v8.0).
2- macro to do mov immediate of 64b values to register.
3- there are misc cleanups in comments and uniform usage of register.
4- GICv3 system interface enabling can only be done at EL3. Specially for the purpose and may be
     future need introduced weak function for platform specific init at EL3.

The patch set is tested on Qemu (boots at EL1) and Broadcom SOC with Cortex A72 ( boots at El3 and drops to EL1) both are arm-v8.0.

Beyond 8.0 more arch specific and core specific changes will be required in future.

Commit:
arch: arm64: refactor for EL3 specific init

Zephyr being an OS is typically expected to run at EL1. Arm core
can reset to EL3 which typically requires a firmware to run at EL3
and drop control to lower EL. In that case EL3 init is done by the
firmware allowing the lower EL software to have necessary control.

If Zephyr is entered at EL3 and it is desired to run at EL1, which
is indicated by 'CONFIG_SWITCH_TO_EL1', then Zephyr is responsible
for doing required EL3 initializations to allow lower EL necessary
control.

The entry sequence is modified to have control flow under single
'switch_el'.

Provisions added by giving weak funcions to do platform specific
init from EL3.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

Commit:
arch: arm64: macro for mov immediate

Single mov instruction can not be used to move non-zero
64b immediate value to the 64b register.
Implement macro to generate mov/ movk and movz sequences
depending on immediate value width.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

carlocaione commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -44,60 +49,89 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
-	mrs	x0, scr_el3
-	orr	x0, x0, #(SCR_EL3_IRQ | SCR_EL3_FIQ | SCR_EL3_EA)
-	msr	scr_el3, x0
+	mrs	x1, scr_el3
```
Why are you avoiding to use the `x0` register?

sandeepbrcm replied:
I kept on changing that couple of times and just thought will do as per review. Intention was to have x0 reserved for special purpose eg: provision to get argument passed from previous stage (if any). But If that happens that can be handled anyway.
  So now its more of choice. can fall back to x0.

carlocaione replied:
Yeah, that's a good thought but I would probably switch back to `x0` for this PR just to keep the diff size as minimal as possible.

sandeepbrcm replied:
ACK .  I will push the new version with uniform usage of registers starting with x0 , x0 ...

carlocaione commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -44,60 +49,89 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
+	mov	x2, #(SCTLR_I_BIT | SCTLR_A_BIT | SCTLR_SA_BIT)
+	mrs	x1, sctlr_el3
+	orr	x1, x1, x2
+	msr	sctlr_el3, x1
```
Same here, reusing the same registers helps in keeping the diff smaller.

sandeepbrcm replied:
ACK

carlocaione commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -118,20 +152,21 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+	 * alignment checks.
 	 */
-	mov	x1, #(SCTLR_I_BIT | SCTLR_A_BIT | SCTLR_SA_BIT)
+	mov	x1, #(SCTLR_I_BIT | SCTLR_SA_BIT | SCTLR_A_BIT)
```
no diff?

sandeepbrcm replied:
ACK.

Commit:
arch: arm64: use callee saved reg to stash

Use calee saved register to preserve value accoss sequence.
Procedure calls are mandated to follow ABI spec and preserve
x19 to x29.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

sandeepbrcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -44,18 +49,65 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
 
-#ifdef CONFIG_SWITCH_TO_EL1
+	/* Setup vector table */
+	adr	x9, _vector_table
```
Typo: I will change x9 to x19 either in another commit in this PR or update in next version.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -44,18 +53,65 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
+	orr     x0, x0, #(ACTLR_EL3_CPUACTLR | ACTLR_EL3_CPUECTLR)
+	msr	actlr_el3, x0
+
+	/* Initialize sctlr_el1 to reset value */
+	mrs     x0, sctlr_el1
+	mov_imm	x1, SCTLR_EL1_RES1
+	orr	x0, x0, x1
+	msr	sctlr_el1, x0
```
I think this can be moved under "1:". 2 reasons for that: 1) consistency: At each EL jump label(3/2/1), we initialize `sctlr_elx` and later set few bits for enabling instruction cache, alignment checks etc. 2) Assuming we directly jump to label1, we need `sctlr_el1` initialization in place.

sandeepbrcm replied:
In standard practice ( ref: tf-a) EL3 init takes care of reset value initialization for EL1 when PE was reset to EL3. 
2) Directly jump to label1:  Implies either PE was reset to EL1 Or  another boot stage ran in EL3 before Zephyr entry.  In both the case 'RES1' setting is not necessary in EL1.

Abhishek-brcm replied:
Okay understood, EL3 is responsible for initializing `sctlr_el1` to `RES1`.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -65,39 +121,21 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
-	 * alignment checks and disable speculative loads.
+	 * alignment checks.
 	 */
 	mov	x1, #(SCTLR_I_BIT | SCTLR_A_BIT | SCTLR_SA_BIT)
```
Based on the experiments that I did earlier, in conjunction with enabling MMU,  we need to disable data access alignment checks by clearing `SCTLR_A_BIT` to enable unaligned access to Normal memory.

sandeepbrcm replied:
Sure. But I thought It will be too much functionality change in one commit. Since you brought this up I will add a small change on top.

sandeepbrcm replied:
update: while I was at it I see I could actually set SCTLR_EL1 to 'RES1' in EL3 instead of doing a read-modify-write , which is wrong in el3 context anyway as core reset to EL3. That way as per armv8.0 SCTLR.A is '0'.  So default behavior is 'Unaligned access allowed to normal memory" . If anyone wants to trap or disable unaligned access for some reason then there can be a commit to set SCTLR.A.

Abhishek-brcm replied:
I see the changes look fine now, `SCTLR_EL1` is directly set to `RES1` and `SCTLR.A` is not getting set in `SCTLR_EL1`.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -65,39 +121,21 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 2:
 	/* Initialize VBAR */
-	msr	vbar_el2, x0
+	msr	vbar_el2, x19
```
"disable speculative loads" can be removed from `sctlr_el2` setting as well. (@ line 150 - mentioning here as not able to put comment there). Although it's debatable if we want to support `EL2` at all. I think `EL1` and `EL3` support is good enough to avoid `EL2` intricacies.

sandeepbrcm replied:
I will change the comment in 'EL2' section as well. We can not support Zephyr in EL3/EL2 as of now, so I ignore the EL2 part.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -19,6 +19,15 @@
+ */
+WTEXT(z_arch_el3_plat_init)
+SECTION_FUNC(TEXT,z_arch_el3_plat_init)
+ret
```
Most EL3 settings should be common between Armv8-A SoCs as done below.. Why is this needed?

sandeepbrcm replied:
As I mentioned in PR description currently GICv3 init is one thing that is qualifying as platform specific. Also few ACTLR init must be done as soon as core is reset as per arm recommendation. Those are indeed platform specific. And some of them are core specific. This is only based on v8.0 analysis.  Beyond that even more such plat_specific inits are foreseen.

carlescufi commented:
@Abhishek-brcm are you happy with this to be merged as-is?

Abhishek-brcm commented:
> @Abhishek-brcm are you happy with this to be merged as-is?

@carlescufi, Comments I had are addressed, looks fine to be merged.
